tools/wavetool.py

```diff
@@ -0,0 +1,157 @@
+        print('Reference wave path does not exist: %s' % ref_path)
+        sys.exit(1)
+    if comparison_mode == 'HWEchoRef':
+        compare_wav_hw_echo_ref(wav_path, ref_path)
+
+def save_wave(wave_data, path, **wave_params):
+    supported_format = ['S8', 'S16', 'S32', 'F32']
+    wave_path = path
+    if path == '.' or not path.endswith('wav'):
+        wave_name = wave_params['type'] + '_' + wave_params['channel'] + 'ch_' + wave_params['freq'] + 'Hz_' \
+        + wave_params['sample_rate'] + '_' + wave_params['format'] + '.wav'
+        wave_path = path + '/'  + wave_name
+    wave_format = wave_params['format'].upper()
+    if wave_format not in supported_format:
+        print('Unsupported wave format: %s' % wave_format)
+        sys.exit(1)
+    if wave_format == 'S8':
+        wave_data = (np.iinfo(np.int8).max * wave_data).astype(np.int8)
+    if wave_format == 'S16':
+        wave_data = (np.iinfo(np.int16).max * wave_data).astype(np.int16)
+    if wave_format == 'S32':
+        wave_data = (np.iinfo(np.int32).max * wave_data).astype(np.int32)
+    try:
+        wavefile.write(wave_path, int(wave_params['sample_rate']), wave_data)
+    except:
+        print("Path specified not valid: %s" % wave_path)
+
+# remove zeros in two sides
+def trim_wave(wave):
+    wave_mono = wave[:,0]
+    left_idx = 0
+    right_idx = wave_mono.shape[0] - 1
+    while True:
+        if wave_mono[left_idx] != 0:
+            break
+        left_idx = left_idx + 1
+    while True:
+        if wave_mono[right_idx] != 0:
+            break
+        right_idx = right_idx - 1
+    return wave[left_idx:right_idx,:], left_idx
+
+# float point binary comparison is not supported, and will not be supported
+# check HW echo referece wave
+def compare_wav_hw_echo_ref(wave_path, ref_path):
+    fs_wav, wave = wavefile.read(wave_path)
+    fs_ref, ref_wav = wavefile.read(ref_path)
+    if fs_wav != fs_ref:
+        print('Can not compare wave with different sample rate')
+        sys.exit(1)
+    trimed_ref_wave, _ = trim_wave(ref_wav)
+    # compare the first two channel
+    trimed_wave_ch_0_1, delay1 = trim_wave(wave[:,0:2])
+    trimed_ref_wave = trimed_ref_wave[0:trimed_wave_ch_0_1.shape[0],:]
+    compare_result0_1 = np.array_equal(trimed_ref_wave, trimed_wave_ch_0_1)
+    # compare the second two channel
+    trimed_wave_ch_2_3, delay2 = trim_wave(wave[:,2:4])
+    trimed_ref_wave = trimed_ref_wave[0:trimed_wave_ch_2_3.shape[0],:]
+    compare_result2_3 = np.array_equal(trimed_ref_wave, trimed_wave_ch_2_3)
+
+    hw_echo_delay = ((delay2 - delay1) / 48000 * 1000)
+    print("delay of hardware echo reference is %0.3fms" % hw_echo_delay)
+    if compare_result0_1 and compare_result2_3 and hw_echo_delay < 5.:
+        print('Recorded wave is binary same as reference wave')
+        print("Wave comparison result: PASSED")
+    else:
+        print('Recorded wave is not binary same as reference wave')
+        print("Wave comparison result: FAILED")
```

Comment:
Use `logging.info()` so this can easily be silenced or redirect later (don't add any silencing or redirection feature yet)

Use an exit code so the caller does not have to grep

Reply:
thanks, will do